mysql/infoschema.go

```diff
@@ -0,0 +1,357 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/lib/pq"
+)
+
+// ProcessInfoSchema performs schema conversion for source database
+// 'db'. We assume that the source database supports information
+// schema tables. These tables are a broadly supported ANSI standard,
+// and we use them to obtain source database's schema information.
+func ProcessInfoSchema(conv *internal.Conv, db *sql.DB) error {
+	tables, err := getTables(db)
+	if err != nil {
+		return err
+	}
+	for _, t := range tables {
+		if err := processTable(conv, db, t); err != nil {
+			return err
+		}
+	}
+	schemaToDDL(conv)
+	conv.AddPrimaryKeys()
+	return nil
+}
+
+// ProcessSqlData performs data conversion for source database
+// 'db'. For each table, we extract data using a "SELECT (colNamesList)" query,
+// convert the data to Spanner data (based on the source and Spanner
+// schemas), and write it to Spanner.  If we can't get/process data
+// for a table, we skip that table and process the remaining tables.
+//
+// Note that the database/sql library has a somewhat complex model for
+// returning data from rows.Scan. We pass *sql.RawBytes to rows.scan.
+// RawBytes is a byte slice and values can be easily converted to string.
+
+func ProcessSqlData(conv *internal.Conv, db *sql.DB) {
+	tables, err := getTables(db)
+	if err != nil {
+		conv.Unexpected(fmt.Sprintf("Couldn't get list of table: %s", err))
+		return
+	}
+	for _, t := range tables {
+		srcTable := buildTableName(t.schema, t.name)
+		srcSchema, ok1 := conv.SrcSchema[srcTable]
+		if !ok1 {
+			conv.Stats.BadRows[srcTable] += conv.Stats.Rows[srcTable]
+			conv.Unexpected(fmt.Sprintf("Can't get schemas for table %s", srcTable))
+			continue
+		}
+		srcCols := srcSchema.ColNames
+		if len(srcCols) == 0 {
+			conv.Unexpected(fmt.Sprintf("Couldn't get source columns for table %s ", t.name))
+			continue
+		}
+		colNameList := buildColNameList(srcSchema, srcCols)
+		// MYSQL schema and name can be arbitrary strings. Ideally we would pass schema/name
+		// as a query parameter,but MYSQL doesn't support this. So we quote it instead.
+		q := fmt.Sprintf(`SELECT %s FROM %s.%s;`, colNameList, t.schema, t.name)
```

Comment:
why not just use SELECT * here?

Reply:
https://github.com/searceinc/harbourbridge/blob/3f1db05b245625dec6e9cbc3ff075e4e3f262783/mysql/infoschema.go#L123-L141

To support spatial data types.